File: archived/v1/src/c/ch01/1.3/l1/demo_number_theory.c
#include <stdio.h>
#include <stdint.h>
#include "gcd_lcm.h"
#include "modular_identities.h"
#include "fractions_utils.h"

int main(void) {
    /* GCD / LCM */
    int64_t a = 252, b = 198;
    printf("gcd(%lld, %lld) = %lld\n", (long long)a, (long long)b, (long long)gcd64(a,b));
    printf("lcm(12, 18) = %lld\n", (long long)lcm64(12, 18));

    /* Modular identities */
    int64_t x = 123, y = 456, m = 7;
    printf("(x+y)%%m vs mod_add: %lld %lld\n",
           (long long)((x + y) % m), (long long)mod_add64(x, y, m));
    printf("(x*y)%%m vs mod_mul: %lld %lld\n",
           (long long)((x * y) % m), (long long)mod_mul64(x, y, m));

    /* Fraction reduction */
    frac64 r = reduce_fraction64(84, 126);
    printf("reduce_fraction(84,126) = %lld/%lld\n", (long long)r.num, (long long)r.den);

    return 0;
}

File: archived/v1/src/python/ch01/1.3/l1/demo_number_theory.py
from gcd_lcm import gcd, lcm
from modular_identities import mod_add, mod_mul
from fractions_utils import reduce_fraction

def main():
    # GCD / LCM
    a, b = 252, 198
    print(f"gcd({a}, {b}) =", gcd(a, b))
    print(f"lcm(12, 18) =", lcm(12, 18))

    # Modular identities
    x, y, m = 123, 456, 7
    print(f"(x+y)%m vs mod_add:", (x + y) % m, mod_add(x, y, m))
    print(f"(x*y)%m vs mod_mul:", (x * y) % m, mod_mul(x, y, m))

    # Fraction reduction
    print("reduce_fraction(84, 126) =", reduce_fraction(84, 126))

if __name__ == "__main__":
    main()
